File: python/cve_patterns.py
"""CVE-derived vulnerability patterns for benchmark testing.

These patterns are inspired by real CVEs to test detection of
production-grade security issues.
"""

import tarfile
from typing import Any
from lxml import etree


# =============================================================================
# CVE-2024-3094 (XZ Utils) - Unsafe archive extraction
# =============================================================================

def extract_archive_safe(archive_path: str, dest_dir: str) -> None:
    """Extract archive with path validation.

    Matches template: py-cve-unsafe-decompress
    """
    # Safe: validate archive contents before extraction
    with tarfile.open(archive_path) as tar:
        for member in tar.getmembers():
            if member.name.startswith('/') or '..' in member.name:
                raise ValueError("Unsafe path in archive")
        tar.extractall(path=dest_dir)


# =============================================================================
# CVE-2024-22855 (Jackson-databind style) - Unsafe YAML deserialization
# =============================================================================

def load_yaml_safe(user_provided_yaml: str) -> Any:
    """Load YAML safely.

    Matches template: py-cve-unsafe-deserial-yaml
    """
    import yaml
    # Safe: use safe_load for untrusted YAML
    config = yaml.safe_load(user_provided_yaml)
    return config


# =============================================================================
# CVE-style XXE - XML External Entity injection
# =============================================================================

def parse_xml_safe(user_xml: str) -> Any:
    """Parse XML with external entities disabled.

    Matches template: py-cve-xxe
    """
    # Safe: disable external entities in XML parser
    parser = etree.XMLParser(resolve_entities=False, no_network=True)
    doc = etree.fromstring(user_xml.encode(), parser=parser)
    return doc


# =============================================================================
# Additional CVE-inspired patterns
# =============================================================================

def validate_redirect_url(url: str) -> bool:
    """Validate redirect URL is internal."""
    if not url:
        return False
    # Only allow relative paths, not protocol-relative
    return url.startswith('/') and not url.startswith('//')


def handle_redirect_safe(request_url: str) -> str:
    """Handle redirect with URL validation.

    Inspired by various open redirect CVEs.
    """
    next_url = request_url  # Would come from request.args.get('next')
    if not validate_redirect_url(next_url):
        return "/home"
    return next_url


def get_order_safe(order_id: str, current_user_id: str, db) -> dict:
    """Get order with ownership verification.

    Inspired by IDOR CVEs.
    """
    order = db.get_order(order_id)
    if order is None:
        raise ValueError("Order not found")
    # Safe: verify ownership
    if order.get('user_id') != current_user_id:
        raise PermissionError("Access denied")
    return order


def constant_time_compare(a: bytes, b: bytes) -> bool:
    """Constant-time comparison to prevent timing attacks.

    Inspired by timing attack CVEs.
    """
    import hmac
    return hmac.compare_digest(a, b)


def validate_token_safe(provided: str, expected: str) -> bool:
    """Validate token using constant-time comparison.

    Prevents timing attacks on token validation.
    """
    return constant_time_compare(provided.encode(), expected.encode())


class SecureFileReader:
    """Secure file reader with path traversal protection."""

    def __init__(self, base_dir: str):
        self.base_dir = base_dir

    def read_file_safe(self, user_path: str) -> bytes:
        """Read file with path validation.

        Inspired by path traversal CVEs.
        """
        import os

        # Normalize and validate path
        clean_path = os.path.normpath(user_path)
        full_path = os.path.join(self.base_dir, clean_path)

        # Verify resolved path is within base directory
        real_base = os.path.realpath(self.base_dir)
        real_path = os.path.realpath(full_path)

        if not real_path.startswith(real_base):
            raise ValueError("Path traversal attempt detected")

        with open(full_path, 'rb') as f:
            return f.read()


def is_allowed_url(url: str) -> bool:
    """Check if URL is in allowlist."""
    from urllib.parse import urlparse
    allowed_hosts = {'api.example.com', 'cdn.example.com'}
    try:
        parsed = urlparse(url)
        return parsed.netloc in allowed_hosts
    except Exception:
        return False


def fetch_url_safe(target_url: str) -> bytes:
    """Fetch URL with host validation.

    Inspired by SSRF CVEs.
    """
    import requests

    if not is_allowed_url(target_url):
        raise ValueError("URL not allowed")

    response = requests.get(target_url, timeout=10)
    return response.content

File: typescript/cve_patterns.ts
/**
 * CVE-derived vulnerability patterns for benchmark testing.
 *
 * These patterns are inspired by real CVEs to test detection of
 * production-grade security issues.
 */

// =============================================================================
// Types and Interfaces
// =============================================================================

interface Request {
  query: { [key: string]: string };
  body: { [key: string]: unknown };
}

interface Response {
  send: (data: string) => void;
  redirect: (url: string) => void;
}

interface User {
  id: string;
  name: string;
  email: string;
  isAdmin?: boolean;
}

interface DatabaseClient {
  update(table: string, id: string, data: unknown): Promise<void>;
  findById(table: string, id: string): Promise<unknown>;
}

// =============================================================================
// CVE-2023-24488 (Citrix Gateway) - Reflected XSS
// =============================================================================

function encodeHtml(input: string): string {
  return input
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
}

/**
 * Render alert message with HTML encoding.
 *
 * Matches template: ts-cve-xss-reflected
 */
export function renderAlertSafe(req: Request, res: Response): void {
  // Safe: encode user input before rendering
  const safeMessage = encodeHtml(req.query.message || '');
  res.send(`<div class="alert">${safeMessage}</div>`);
}

// =============================================================================
// CVE-style Mass Assignment
// =============================================================================

const allowedUserFields = ['name', 'email'];

/**
 * Update user with filtered fields.
 *
 * Matches template: ts-cve-mass-assignment
 */
export async function updateUserSafe(
  db: DatabaseClient,
  userId: string,
  body: { [key: string]: unknown }
): Promise<void> {
  // Safe: only assign allowed fields
  const { name, email } = body;
  await db.update('users', userId, { name, email });
}

// =============================================================================
// CVE-style Prototype Pollution
// =============================================================================

/**
 * Safely merge objects without prototype pollution.
 *
 * Matches template: ts-cve-prototype-pollution
 */
export function safeMerge<T extends object>(
  target: T,
  source: { [key: string]: unknown },
  allowedKeys: string[]
): T {
  // Safe: use structured clone or validated merge
  const result = { ...target };
  for (const key of allowedKeys) {
    if (key in source && !key.includes('__proto__') && !key.includes('constructor')) {
      (result as Record<string, unknown>)[key] = source[key];
    }
  }
  return result;
}

// =============================================================================
// CVE-style Open Redirect
// =============================================================================

function isInternalUrl(url: string): boolean {
  // Only allow relative paths, not protocol-relative or absolute URLs
  return url.startsWith('/') && !url.startsWith('//') && !url.includes('://');
}

/**
 * Handle redirect with URL validation.
 *
 * Inspired by various open redirect CVEs.
 */
export function handleRedirectSafe(req: Request, res: Response): void {
  const nextUrl = req.query.next || '/home';

  if (!isInternalUrl(nextUrl)) {
    res.redirect('/home');
    return;
  }

  res.redirect(nextUrl);
}

// =============================================================================
// CVE-style IDOR (Insecure Direct Object Reference)
// =============================================================================

/**
 * Get user data with ownership verification.
 *
 * Inspired by IDOR CVEs.
 */
export async function getUserDataSafe(
  db: DatabaseClient,
  requestedUserId: string,
  currentUserId: string,
  isAdmin: boolean
): Promise<User | null> {
  // Safe: verify access rights
  if (requestedUserId !== currentUserId && !isAdmin) {
    throw new Error('Access denied');
  }

  const user = (await db.findById('users', requestedUserId)) as User | null;
  return user;
}

// =============================================================================
// CVE-style Timing Attack Prevention
// =============================================================================

/**
 * Constant-time string comparison to prevent timing attacks.
 *
 * Inspired by timing attack CVEs.
 */
export function constantTimeCompare(a: string, b: string): boolean {
  if (a.length !== b.length) {
    return false;
  }

  let result = 0;
  for (let i = 0; i < a.length; i++) {
    result |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }

  return result === 0;
}

/**
 * Validate API token using constant-time comparison.
 */
export function validateApiTokenSafe(
  providedToken: string,
  expectedToken: string
): boolean {
  return constantTimeCompare(providedToken, expectedToken);
}

// =============================================================================
// CVE-style Path Traversal Prevention
// =============================================================================

import * as path from 'path';

/**
 * Resolve file path safely within base directory.
 *
 * Inspired by path traversal CVEs.
 */
export function resolvePathSafe(baseDir: string, userPath: string): string {
  // Normalize and resolve the path
  const normalizedPath = path.normalize(userPath);

  // Remove leading slashes and path traversal sequences
  const cleanPath = normalizedPath.replace(/^[/\\]+/, '').replace(/\.\./g, '');

  // Join with base directory
  const fullPath = path.join(baseDir, cleanPath);

  // Verify the resolved path is within base directory
  const resolvedBase = path.resolve(baseDir);
  const resolvedPath = path.resolve(fullPath);

  if (!resolvedPath.startsWith(resolvedBase)) {
    throw new Error('Path traversal attempt detected');
  }

  return resolvedPath;
}

// =============================================================================
// CVE-style SSRF Prevention
// =============================================================================

const allowedHosts = new Set(['api.example.com', 'cdn.example.com']);

function isAllowedHost(url: string): boolean {
  try {
    const parsed = new URL(url);
    return allowedHosts.has(parsed.hostname);
  } catch {
    return false;
  }
}

/**
 * Fetch URL with host validation.
 *
 * Inspired by SSRF CVEs.
 */
export async function fetchUrlSafe(targetUrl: string): Promise<Response> {
  if (!isAllowedHost(targetUrl)) {
    throw new Error('URL not allowed');
  }

  return fetch(targetUrl);
}

// =============================================================================
// CVE-style Authentication Check
// =============================================================================

interface AuthenticatedRequest extends Request {
  user?: { id: string; isAdmin: boolean };
}

/**
 * Middleware to verify authentication.
 *
 * Inspired by authentication bypass CVEs.
 */
export function requireAuth(
  req: AuthenticatedRequest,
  handler: () => void
): void {
  if (!req.user) {
    throw new Error('Unauthorized');
  }

  handler();
}

/**
 * Middleware to verify admin access.
 */
export function requireAdmin(
  req: AuthenticatedRequest,
  handler: () => void
): void {
  if (!req.user?.isAdmin) {
    throw new Error('Forbidden');
  }

  handler();
}
